BUG: DatetimeIndex.shift(freq=...) raises near DST boundary

xref #5694 #8531 (?)
xref #8817 

This is presumably caused by the fact that pytz time zones internalize the offset of the current time.

``` python
In [1]: import pandas as pd

In [2]: idx = pd.date_range('2013-11-03', tz='America/Chicago',
   ...:                     periods=6, freq='H')

In [3]: pd.Series(index=idx)
Out[3]: 
2013-11-03 00:00:00-05:00   NaN
2013-11-03 01:00:00-05:00   NaN
2013-11-03 01:00:00-06:00   NaN
2013-11-03 02:00:00-06:00   NaN
2013-11-03 03:00:00-06:00   NaN
2013-11-03 04:00:00-06:00   NaN
Freq: H, dtype: float64

In [4]: pd.Series(index=idx).shift(freq='H')
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-4-19ee418c9aa1> in <module>()
----> 1 pd.Series(index=idx).shift(freq='H')

/Users/afni/homebrew/Cellar/python/2.7.5/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pandas/core/generic.pyc in shift(self, periods, freq, axis, **kwds)
   3290             new_data = self._data.shift(periods=periods, axis=block_axis)
   3291         else:
-> 3292             return self.tshift(periods, freq, **kwds)
   3293 
   3294         return self._constructor(new_data).__finalize__(self)

/Users/afni/homebrew/Cellar/python/2.7.5/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pandas/core/generic.pyc in tshift(self, periods, freq, axis, **kwds)
   3386         else:
   3387             new_data = self._data.copy()
-> 3388             new_data.axes[block_axis] = index.shift(periods, offset)
   3389 
   3390         return self._constructor(new_data).__finalize__(self)

/Users/afni/homebrew/Cellar/python/2.7.5/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pandas/tseries/index.pyc in shift(self, n, freq)
    855         end = self[-1] + n * self.offset
    856         return DatetimeIndex(start=start, end=end, freq=self.offset,
--> 857                              name=self.name, tz=self.tz)
    858 
    859     def repeat(self, repeats, axis=None):

/Users/afni/homebrew/Cellar/python/2.7.5/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pandas/tseries/index.pyc in __new__(cls, data, freq, start, end, periods, copy, name, tz, verify_integrity, normalize, closed, **kwds)
    204             return cls._generate(start, end, periods, name, freq,
    205                                  tz=tz, normalize=normalize, closed=closed,
--> 206                                  infer_dst=infer_dst)
    207 
    208         if not isinstance(data, (np.ndarray, ABCSeries)):

/Users/afni/homebrew/Cellar/python/2.7.5/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pandas/tseries/index.pyc in _generate(cls, start, end, periods, name, offset, tz, normalize, infer_dst, closed)
    369         if tz is not None and inferred_tz is not None:
    370             if not inferred_tz == tz:
--> 371                 raise AssertionError("Inferred time zone not equal to passed "
    372                                      "time zone")
    373 

AssertionError: Inferred time zone not equal to passed time zone
```

Playing with `dateutil` and `pytz` timezones makes me despair of repairing that assertion to be anything sane.  Is it actually needed, or can we just turn it into a warning?

Observations:
- two `pytz.DstTzInfo`s of the same zoneinfo zone that have been `.normalize()`ed to different times may compare different
- a `pytz.*TzInfo` has a `zone` member containing the name of the zoneinfo zone from which it was constructed
- a `dateutil.tzfile` has no public members beyond the `tzinfo` API, which is insufficient to tell whether two time zones are equal
- a `pytz` time zone and a `dateutil` time zone appear to compare different under all circumstances, regardless of whether they represent the same time zone
- the various `dateutil` time zone classes have no common base class within the `dateutil` package
- `pytz` zones constructed from different names for the same zoneinfo zone (e.g. `UTC` and `Etc/UTC`) compare different

Ugh.

from #8817

```
import pandas as pd
import pytz
import datetime

dt = datetime.datetime(2014, 11, 14, 0)
dt_est = pytz.timezone('EST').localize(dt)
s = pd.Series(data=[1], index=[dt_est])

s.shift(0, freq='h')  # 2014-11-14 00:00:00-05:00 (seems okay) 
s.shift(-1, freq='h')  # 2014-11-13 18:00:00-05:00 (expected 2014-11-13 23:00:00)
s.shift(1, freq='h')  # 2014-11-13 20:00:00-05:00 (expected 2014-11-14 01:00:00)

s.shift(-1, freq='s')  # 2014-11-13 18:59:59-05:00 (same with other freq)
```

--- a/doc/source/whatsnew/v0.23.2.txt
+++ b/doc/source/whatsnew/v0.23.2.txt
@@ -79,6 +79,11 @@ Bug Fixes
 **Timezones**
 - Bug in :class:`Timestamp` and :class:`DatetimeIndex` where passing a :class:`Timestamp` localized after a DST transition would return a datetime before the DST transition (:issue:`20854`)
 - Bug in comparing :class:`DataFrame`s with tz-aware :class:`DatetimeIndex` columns with a DST transition that raised a ``KeyError`` (:issue:`19970`)
+- Bug in :meth:`DatetimeIndex.shift` where an ``AssertionError`` would raise when shifting across DST (:issue:`8616`)
+- Bug in :class:`Timestamp` constructor where passing an invalid timezone offset designator (``Z``) would not raise (:issue:`8910`)
+- Bug in :meth:`Timestamp.replace` where replacing at a DST boundary would retain an incorrect offset (:issue:`7825`)
+- Bug in :meth:`DatetimeIndex.reindex` where reindexing a tz-naive and tz-aware :class:`DatetimeIndex` would not raise (:issue:`8306`)
+- Bug in :meth:`DatetimeIndex.resample` where downsampling across a DST boundary would raise (:issue:`8531`)
 
 **Other**
 

--- a/doc/source/whatsnew/v0.23.2.txt
+++ b/doc/source/whatsnew/v0.23.2.txt
@@ -80,10 +80,10 @@ Bug Fixes
 - Bug in :class:`Timestamp` and :class:`DatetimeIndex` where passing a :class:`Timestamp` localized after a DST transition would return a datetime before the DST transition (:issue:`20854`)
 - Bug in comparing :class:`DataFrame`s with tz-aware :class:`DatetimeIndex` columns with a DST transition that raised a ``KeyError`` (:issue:`19970`)
 - Bug in :meth:`DatetimeIndex.shift` where an ``AssertionError`` would raise when shifting across DST (:issue:`8616`)
-- Bug in :class:`Timestamp` constructor where passing an invalid timezone offset designator (``Z``) would not raise (:issue:`8910`)
+- Bug in :class:`Timestamp` constructor where passing an invalid timezone offset designator (``Z``) would not raise a ``ValueError``(:issue:`8910`)
 - Bug in :meth:`Timestamp.replace` where replacing at a DST boundary would retain an incorrect offset (:issue:`7825`)
-- Bug in :meth:`DatetimeIndex.reindex` where reindexing a tz-naive and tz-aware :class:`DatetimeIndex` would not raise (:issue:`8306`)
-- Bug in :meth:`DatetimeIndex.resample` where downsampling across a DST boundary would raise (:issue:`8531`)
+- Bug in :meth:`DatetimeIndex.reindex` when reindexing a tz-naive and tz-aware :class:`DatetimeIndex` (:issue:`8306`)
+- Bug in :meth:`DatetimeIndex.resample` when downsampling across a DST boundary (:issue:`8531`)
 
 **Other**
 

--- a/pandas/tests/indexes/datetimes/test_arithmetic.py
+++ b/pandas/tests/indexes/datetimes/test_arithmetic.py
@@ -476,6 +476,28 @@ class TestDatetimeIndexArithmetic(object):
         result = dr_tz.shift(1, '10T')
         assert result.tz == dr_tz.tz
 
+    def test_dti_shift_across_dst(self):
+        # GH 8616
+        idx = date_range('2013-11-03', tz='America/Chicago',
+                         periods=7, freq='H')
+        s = Series(index=idx[:-1])
+        result = s.shift(freq='H')
+        expected = Series(index=idx[1:])
+        tm.assert_series_equal(result, expected)
+
+    @pytest.mark.parametrize('shift, result',[
+        [0, '2014-11-14 00:00:00'],
+        [-1, '2014-11-13 23:00:00',
+        [1, '2014-11-14 01:00:00']]])
+    def test_dti_shift_near_midnight(self, shift, result):
+        # GH 8616
+        dt = datetime(2014, 11, 14, 0)
+        dt_est = pytz.timezone('EST').localize(dt)
+        s = Series(data=[1], index=[dt_est])
+        result = s.shift(shift, freq='H')
+        expected = Series(1, index=DatetimeIndex([result], tz='US/Eastern'))
+        tm.assert_series_equal(result, expected) 
+
     # -------------------------------------------------------------
     # Binary operations DatetimeIndex and timedelta-like
 

--- a/pandas/tests/indexes/datetimes/test_arithmetic.py
+++ b/pandas/tests/indexes/datetimes/test_arithmetic.py
@@ -4,7 +4,7 @@ from datetime import datetime, timedelta
 import operator
 
 import pytest
-
+import pytz
 import numpy as np
 
 import pandas as pd
@@ -485,18 +485,18 @@ class TestDatetimeIndexArithmetic(object):
         expected = Series(index=idx[1:])
         tm.assert_series_equal(result, expected)
 
-    @pytest.mark.parametrize('shift, result',[
+    @pytest.mark.parametrize('shift, result_time', [
         [0, '2014-11-14 00:00:00'],
-        [-1, '2014-11-13 23:00:00',
-        [1, '2014-11-14 01:00:00']]])
-    def test_dti_shift_near_midnight(self, shift, result):
+        [-1, '2014-11-13 23:00:00'],
+        [1, '2014-11-14 01:00:00']])
+    def test_dti_shift_near_midnight(self, shift, result_time):
         # GH 8616
         dt = datetime(2014, 11, 14, 0)
         dt_est = pytz.timezone('EST').localize(dt)
         s = Series(data=[1], index=[dt_est])
         result = s.shift(shift, freq='H')
-        expected = Series(1, index=DatetimeIndex([result], tz='US/Eastern'))
-        tm.assert_series_equal(result, expected) 
+        expected = Series(1, index=DatetimeIndex([result_time], tz='EST'))
+        tm.assert_series_equal(result, expected)
 
     # -------------------------------------------------------------
     # Binary operations DatetimeIndex and timedelta-like

--- a/pandas/tests/scalar/timestamp/test_timestamp.py
+++ b/pandas/tests/scalar/timestamp/test_timestamp.py
@@ -420,6 +420,12 @@ class TestTimestampConstructors(object):
         expected = expected + Timedelta(nanoseconds=1)
         assert result == expected
 
+    @pytest.mark.parametrize('z', ['Z0', 'Z00'])
+    def test_constructor_invalid_Z0_isostring(self, z):
+        # GH 8910
+        with pytest.raises(ValueError):
+            Timestamp('2014-11-02 01:00{}'.format(z))
+
     @pytest.mark.parametrize('arg', ['year', 'month', 'day', 'hour', 'minute',
                                      'second', 'microsecond', 'nanosecond'])
     def test_invalid_date_kwarg_with_string_input(self, arg):

--- a/pandas/tests/scalar/timestamp/test_unary_ops.py
+++ b/pandas/tests/scalar/timestamp/test_unary_ops.py
@@ -238,6 +238,13 @@ class TestTimestampUnaryOps(object):
         ts2b = normalize(ts2)
         assert ts2 == ts2b
 
+    def test_replace_dst_border(self):
+        # Gh 7825
+        t = Timestamp('2013-11-3', tz='America/Chicago')
+        result = t.replace(hour=3)
+        expected = Timestamp('2013-11-3 03:00:00', tz='America/Chicago')
+        assert result == expected
+
     # --------------------------------------------------------------
 
     @td.skip_if_windows

--- a/pandas/tests/series/indexing/test_alter_index.py
+++ b/pandas/tests/series/indexing/test_alter_index.py
@@ -453,6 +453,15 @@ def test_reindex_fill_value():
     assert_series_equal(result, expected)
 
 
+def test_reindex_datetimeindexes_tz_naive_and_aware():
+    # GH 8306
+    idx = date_range('20131101', tz='America/Chicago', periods=7)
+    newidx = date_range('20131103', periods=10, freq='H')
+    s = Series(range(7), index=idx)
+    with pytest.raises(TypeError):
+        s.reindex(newidx, method='ffill')
+
+
 def test_rename():
     # GH 17407
     s = Series(range(1, 6), index=pd.Index(range(2, 7), name='IntIndex'))

--- a/pandas/tests/test_resample.py
+++ b/pandas/tests/test_resample.py
@@ -2084,6 +2084,17 @@ class TestDatetimeIndex(Base):
                                        freq='D', tz='Europe/Paris')),
             'D Frequency')
 
+    def test_downsample_across_dst(self):
+        # GH 8531
+        tz = pytz.timezone('Europe/Berlin')
+        dt = datetime(2014,10,26)
+        dates = date_range(tz.localize(dt), periods=4, freq='2H')
+        result = Series(5, index=dates).resample('H').mean()
+        expected = Series([5., np.nan] * 3 + [5.],
+                          index=date_range(tz.localize(dt), periods=7,
+                                           freq='H'))
+        tm.assert_frame_equal(result, expected)
+
     def test_resample_with_nat(self):
         # GH 13020
         index = DatetimeIndex([pd.NaT,

--- a/pandas/tests/test_resample.py
+++ b/pandas/tests/test_resample.py
@@ -2087,13 +2087,13 @@ class TestDatetimeIndex(Base):
     def test_downsample_across_dst(self):
         # GH 8531
         tz = pytz.timezone('Europe/Berlin')
-        dt = datetime(2014,10,26)
+        dt = datetime(2014, 10, 26)
         dates = date_range(tz.localize(dt), periods=4, freq='2H')
         result = Series(5, index=dates).resample('H').mean()
         expected = Series([5., np.nan] * 3 + [5.],
                           index=date_range(tz.localize(dt), periods=7,
                                            freq='H'))
-        tm.assert_frame_equal(result, expected)
+        tm.assert_series_equal(result, expected)
 
     def test_resample_with_nat(self):
         # GH 13020

